src/python/pants/cache/restful_artifact_cache.py

```diff
@@ -5,28 +5,157 @@
 import multiprocessing
 import queue
 import threading
+from contextlib import contextmanager
+from dataclasses import dataclass
 
 import requests
 from requests import RequestException
+from urllib3.exceptions import MaxRetryError
+from urllib3.util.retry import Retry
 
 from pants.cache.artifact_cache import ArtifactCache, NonfatalArtifactCacheError, UnreadableArtifact
+from pants.subsystem.subsystem import Subsystem
+from pants.util.memo import memoized_classmethod
+from pants.util.meta import staticproperty
 
 
 logger = logging.getLogger(__name__)
 
-# Reduce the somewhat verbose logging of requests.
-# TODO do this in a central place
-logging.getLogger('requests').setLevel(logging.WARNING)
-
 
+@dataclass(frozen=True)
 class RequestsSession:
-  _session = None
+  """Configuration for the session object used to fetch HTTP(S) artifact cache entries.
+
+  See
+  https://github.com/psf/requests/blob/d2590ee46c0641958b6d4792a206bd5171cb247d/requests/adapters.py#L113
+  for documentation on the HTTPAdapter class, and check out
+  https://urllib3.readthedocs.io/en/latest/advanced-usage.html#customizing-pool-behavior for
+  a description of the connection pooling implementation in urllib3.
+
+  See
+  https://github.com/urllib3/urllib3/blob/f4b36ad045ccfbbfaaadd8e69f9b32c5d81cbd84/src/urllib3/util/retry.py#L29
+  for the Retry class (also from urllib3) which is used here to configure request retries.
+  """
+  max_connection_pools: int
+  max_connections_within_pool: int
+  max_retries: int
+  max_retries_on_connection_errors: int
+  max_retries_on_read_errors: int
+  backoff_factor: float
+  blocking_pool: bool
+  slow_download_timeout_seconds: int
+
+  # Global flag which is set to True if our global connection pool singleton raises a MaxRetryError.
+  _max_retries_exceeded = False
+
+  class Factory(Subsystem):
+    options_scope = 'http-artifact-cache'
+
+    # Maintain a connection pool of max size equaling the larger of the number of available cores,
+    # or the default from the requests package (which is usually 10).
+    _default_pool_size = max(multiprocessing.cpu_count(), requests.adapters.DEFAULT_POOLSIZE)
+    # By default, don't perform any retries.
+    _default_retries = 0
+
+    @classmethod
+    def register_options(cls, register):
+      super().register_options(register)
+      # TODO: Pull the `choices` from the registered log levels in the `logging` module somehow!
+      register('--requests-logging-level', choices=['WARNING', 'INFO', 'DEBUG'],
```

Comment:
Should this be an `Enum`? I'm unclear when to use this vs a list of strings.

Reply:
It should be an `Enum`! I have made the change!